pkg/azure/aks/machine_store.go

```diff
@@ -0,0 +1,253 @@
+		azVMSSVmClient:        computeClientFactory.NewVirtualMachineScaleSetVMsClient(),
+		azAksClient:           containerClientFactory.NewManagedClustersClient(),
+
+		MachineMap:     make(map[string]*VirtualMachineInfo),
+		machineMapLock: &sync.RWMutex{},
+	}
+
+	go func() {
+		err := ms.PopulateMachineStore(ms.context)
+		if err != nil {
+			ms.logger.LogAttrs(ms.context, slog.LevelError, "error populating initial machine store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return ms, nil
+}
+
+func (m *MachineStore) getVmInfoByVmName(vmName string) (*VirtualMachineInfo, error) {
+	m.machineMapLock.RLock()
+	defer m.machineMapLock.RUnlock()
+
+	vmInfo := m.MachineMap[vmName]
+	if vmInfo == nil {
+		return nil, ErrMachineNotFound
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromVmss(ctx context.Context, rgName string, vmssName string, priority MachinePriority, osInfo MachineOperatingSystem) (map[string]*VirtualMachineInfo, error) {
+	vmInfo := make(map[string]*VirtualMachineInfo)
+
+	opts := &armcompute.VirtualMachineScaleSetVMsClientListOptions{
+		Expand: to.StringPtr("instanceView"),
+	}
+	pager := m.azVMSSVmClient.NewListPager(rgName, vmssName, opts)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in VMSS VM Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+
+		for _, v := range nextResult.Value {
+			vmName, err := getMachineName(v)
+			if err != nil {
+				m.logger.Error(err.Error())
+				continue
+			}
+
+			vmInfo[vmName] = &VirtualMachineInfo{
+				Name:            vmName,
+				Region:          to.String(v.Location),
+				OwningVMSS:      vmssName,
+				MachineTypeSku:  to.String(v.SKU.Name),
+				Priority:        priority,
+				OperatingSystem: osInfo,
+			}
+			m.logger.LogAttrs(ctx, slog.LevelDebug, "found machine information", slog.String("machineName", vmName))
+		}
+	}
+
+	return vmInfo, nil
+}
+
+func (m *MachineStore) getVmInfoFromResourceGroup(ctx context.Context, rgName, clusterName string) (map[string]*VirtualMachineInfo, error) {
+	vmInfoMap := make(map[string]*VirtualMachineInfo)
+
+	pager := m.azVMSSClient.NewListPager(rgName, nil)
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			m.logger.LogAttrs(ctx, slog.LevelError, "unable to advance page in VMSS Client", slog.String("err", err.Error()))
+			return nil, ErrPageAdvanceFailure
+		}
+
+		for _, v := range nextResult.Value {
+			vmssName := to.String(v.Name)
+			vmssPriority := getMachineScaleSetPriority(v)
+			vmssOperationSystem := getMachineScaleSetOperatingSystem(v)
+
+			vmInfo, err := m.getVmInfoFromVmss(ctx, rgName, vmssName, vmssPriority, vmssOperationSystem)
+			if err != nil {
+				return nil, err
+			}
+
+			for name, info := range vmInfo {
+				info.OwningCluster = clusterName
+				vmInfoMap[name] = info
+			}
```

Comment:
I'm struggling a bit on if this makes sense within this call, or if you should return a list of `vmssName`'s and have a subsequent call to list vm info. It would require you to alter  https://github.com/grafana/cloudcost-exporter/pull/221/files#diff-e7c4fb1d534ac6235c7771f7207ceeb022c725aca0022efcc21e1990f5a2191fR213-R220 

Some of the benefits:
- Decouple getVmInfoFromResourceGroup from getVmInfoFromVmss
- Process fetching vm info in parallel
- Make the code a bit more readable as `PopulateMachineStore` has the happy path explicitly written out  

How does that sound?

Reply:
This is a good idea. I'll implement

Reply:
Changed to have these listed out individually and wrapped by `PopulateMachineStore`, also allowing the calls to be parallelized with an errgroup.  WDYT?